src/ORM/DataObject.php

```diff
@@ -4638,4 +4643,37 @@ public function findAllRelatedData(array $excludedClasses = []): SS_List
         $service = Injector::inst()->get(RelatedDataService::class);
         return $service->findAll($this, $excludedClasses);
     }
+
+    private function buildValidationResultForDuplicateEntry(DuplicateEntryException $exception): ValidationResult
+    {
+        $key = $exception->getKeyName();
+        $singleName = static::i18n_singular_name();
+        $columns = DataObject::getSchema()->getIndexFields(static::class, $key);
+        $validationResult = ValidationResult::create();
+        if (empty($columns)) {
+            $validationResult->addError(_t(
+                __CLASS__ . '.NO_DUPLICATE',
+                'Cannot create duplicate {type}',
+                ['type' => $singleName]
+            ));
```

Comment:
There's no way to trigger this intentionally - but it _is_ theoretically possible that the key won't match what we expect, or someone throws the exception manually and doesn't set a key.
In those scenarios we should still provide the best validation error we're able to.